File: day5.py
# http://adventofcode.com/2017/day/5

def solve(instructions, part):
    jump = 0
    steps_taken = 0
    local_copy_instructions = list(instructions)
    instruction_count = len(local_copy_instructions)

    while jump < instruction_count and jump >= 0:
        j = local_copy_instructions[jump]
        if j >= 3 and part == 1:
            local_copy_instructions[jump] -= 1
        else:
            local_copy_instructions[jump] += 1
        jump += j
        steps_taken += 1

    return steps_taken

instructions = list(map(int,[x.rstrip() for x in open('day5-input.txt','r').readlines()]))

print(solve(instructions, 0))
print(solve(instructions, 1))

File: day5.go
// http://adventofcode.com/2017/day/5

package main

import (
	"os"
	"bufio"
	"strconv"
	"fmt"
)

func solveDay5(instructions []int64, isPart2 bool) int64 {
	steps := int64(0)
	jump := int64(0)
	localCopyOfInstructions := make([]int64, len(instructions))
	copy(localCopyOfInstructions, instructions)
	instructionCount := int64(len(localCopyOfInstructions))
	for jump < instructionCount && jump >= 0 {
		j := localCopyOfInstructions[jump]
		if j >= 3 && isPart2 {
			localCopyOfInstructions[jump] -= 1
		} else {
			localCopyOfInstructions[jump] += 1
		}
		jump += j
		steps++
	}

	return steps
}

func main() {
	f, err := os.Open("day5-input.txt")
	if err != nil {
		panic(err)
	}
	scanner := bufio.NewScanner(f)

	var instructionList []int64

	for scanner.Scan() {
		jump, _ := strconv.ParseInt(scanner.Text(), 10, 64)
		instructionList = append(instructionList, jump)
	}
	defer f.Close()
	fmt.Println(solveDay5(instructionList, false))
	fmt.Println(solveDay5(instructionList, true))
}